File: fsharp/binary-search/BinarySearch.fs
module BinarySearch

let find input value =
    let rec search lst =
        let left, right =
            match List.length lst with
            | 0 -> [], []
            | 1 -> [], lst
            | n -> List.splitAt (n / 2 - 1) lst

        match left, right with
        | _, (i, x) :: _ when x = value -> Some i
        | l, (_, x) :: _ when x > value -> search l
        | _, (_, x) :: r when x < value -> search r
        | _ -> None

    input |> Array.toList |> List.indexed |> search

File: csharp/binary-search/BinarySearch.cs
public static class BinarySearch
{
    public static int Find(int[] input, int value)
    {
        var lo = 0;
        var hi = input.Length - 1;
        
        while (lo <= hi)
        {
            var mid = (lo + hi) / 2;

            if (input[mid] == value)
                return mid;

            if (input[mid] < value)
                lo = mid + 1;
            else
                hi = mid - 1;
        }

        return -1;
    }
}